Internal _cdf methods with a precomputed CDF

jashan498 opened the pull request:
### Fixes #13830

Added various **_cdf** methods which contain pre-computed CDF for all random variables mentioned by  @normalhuman in the issue. Following are the continuous random variable for which _cdf method has been added:

1. Arcsin
2. Dagnum
3. Erlang
4. Frechet
5. Gamma
6. GammaInverse
7. Kumaraswamy
8. Laplace
9. Logistic
10. Nakagami
11. StudentT
12. UniformSum

jashan498 committed:
_cdf method and corresponding tests added for Arcsin and Dagnum

jashan498 committed:
Added precomputed cdf for following continuous random variables :
Erlang
Frechet
Gamma
GammaInverse
Kumaraswamy
Laplace
Logistic
Nakagami

Also added _cdf function for StudentT and UniformSum, but they require further improvements

jashan498 committed:
Changed outputs of all the above _cdf functions in terms of Piecewise function

jashan498 committed:
Improvised _cdf for
StudentT
UniformSum

normalhuman commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -458,6 +490,8 @@ def test_studentt():
 
     X = StudentT('x', nu)
     assert density(X)(x) == (1 + x**2/nu)**(-nu/2 - 1/2)/(sqrt(nu)*beta(1/2, nu/2))
+    assert cdf(X)(x) == 1/2 + x*gamma(nu/2 + 1/2)*hyper((1/2, nu/2 + 1/2),
```
A problem with these tests (both for cdf and for density) is that they are essentially the same code pasted twice. They don't really check how the formulas __work__, they only check what the formulas __are__.  Try to add a test that involves  
`diff(cdf(X)(x), x)`  compared to `density(X)(x)`. That would make sure that the two things are consistent: the derivative of the CDF is the PDF. 

It's quite possible that `Eq(diff(cdf(X)(x), x), density(X)(x)).simplify()` will fail to simplify to True, because of how complex the formulas are. But perhaps you can do `.subs(x, Rational(random.gauss(0, 1)))` in there, verifying the equality at a randomly chosen rational point. Worth trying. 
  
There is already `test_precomputed_cdf` but it is based on integrating the PDF and we know that won't work well because these PDF can't be integrated easily. But taking the derivative of cdf should be easier.

jashan498 replied:
Should I add this test for every method or just for the ones I have added?

normalhuman replied:
I imagine this could be one test running in a loop over distributions. See how test_precomputed_cdf is written.

jashan498 committed:
added an example for evaluation of cdf of UniformSum distribution with specific values of 'x' and 'n'

jashan498 committed:
Added test_long_precomputed_cdf() function and corrected formula for precomputed cdf of dagum and logistic random variables
test_long_precomputed_cdf() also will keep a check on the consistency of density and cdf functions as they both are related.

jashan498 committed:
Added test for UniformSums cdf and density

normalhuman commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -709,7 +750,34 @@ def test_precomputed_cdf():
+            ]
+    for distr in distribs:
+        for _ in range(5):
+            assert abs((simplify(diff(cdf(distr)(x), x))
```
You probably don't need simplify here, since you are just plugging in a number for floating point evaluation. simplify is slow. Also, you don't even need all of the import random thing: there is `verify_numerically` method which takes care of things. See an example [here](https://github.com/sympy/sympy/pull/13909). Use it with `a=0, b=0, c=1, d=0` and you'll get verification for random numbers between 0 and 1. 

E.g., `assert tn(diff(cdf(distr)(x), x), density(distr)(x), x, a=0, b=0, c=1, d=0)`. It's much cleaner. 

Same for UniformSum cdf below.

normalhuman commented:
You'll need to resolve merge conflicts here, since characteristic_function was added meanwhile. Use git merge as [described here](https://github.com/sympy/sympy/wiki/Development-workflow#synchronize-with-master-sympy-sympy)

jashan498 committed:
Merge branch 'master' into precompute_cdf

jashan498 committed:
Changed random with verify_numerically method

jashan498 committed:
deleted one whitespace

normalhuman commented on sympy/stats/crv_types.py:
```diff
@@ -2603,6 +2719,19 @@ def UniformSum(name, n):
+    >>> cdf(UniformSum("x", 5), evaluate=False)(2).doit()
+    9/40
+
+    In above formof cdf functtion evaluate=False prevents an attempt at premature evaluation of the sum,
```
"formof" is  a typo which I haven't noticed earlier. The second sentence seems unnecessary: we see it was evaluated. This should be shorter. I suggest:

The argument evaluate=False prevents an attempt at evaluation of the sum for general n, before the argument 2 is passed.

Also, docstrings should be wrapped at 72 characters, if I remember correctly (from your previous PR)

jashan498 committed:
some minor changes in docstring

normalhuman reviewed (approved):
+1 The entered CDFs are numerically checked against PDFs (which were in the module earlier) by differentiation, which provides additional confirmation that both CDFs and PDFs are correct.

normalhuman commented on sympy/stats/crv_types.py:
```diff
@@ -2721,16 +2721,17 @@ def UniformSum(name, n):
-    evaluated with doit and without any hypergeometric complications.
+    The argument evaluate=False prevents an attempt at evaluation
+    of the sum for general n, before the argument 2 is passed.
+    Once 2 is passed, the sum is easilyevaluated with doit and without
```
This second sentence has a typo and it should not be here at all, it does not fit the style of docstrings. My earlier comment here was meant that one sentence I suggested (and you put in) was enough. Saying "it works if you do it this way" is not needed, this is what the example already shows. Sorry if this wasn't clear.

normalhuman committed:
Shorten a docstring

This was an unnecessary sentence, a bit too conversational.

gxyd commented:
Welcome to SymPy! @normalhuman :). I hope I'm not too late on that.
